compiler-core/src/package_interface.rs

```diff
@@ -0,0 +1,280 @@
+use std::{collections::HashMap, ops::Deref};
+
+use ecow::EcoString;
+use hexpm::version::Version;
+use serde::Serialize;
+
+#[cfg(test)]
+mod tests;
+
+use crate::{
+    ast::{CustomType, Definition, Function, ModuleConstant},
+    manifest::ordered_map,
+    type_::{Deprecation, Type, TypeVar},
+};
+
+use crate::build::{Module, Package, Target};
+
+#[derive(Serialize, Debug)]
+pub struct PackageInterface {
+    name: EcoString,
+    package_version: Version,
+    gleam_version: Option<EcoString>,
+    #[serde(serialize_with = "ordered_map")]
+    modules: HashMap<EcoString, ModuleInterface>,
+}
+
+/// Holds serialisable data about a `Module`.
+#[derive(Serialize, Debug)]
+pub struct ModuleInterface {
+    documentation: Vec<EcoString>,
+    #[serde(serialize_with = "ordered_map")]
+    types: HashMap<EcoString, TypeDefinitionInterface>,
+    #[serde(serialize_with = "ordered_map")]
+    values: HashMap<EcoString, ValueInterface>,
+}
+
+/// Holds serialisable data about a type definition that appears in a module.
+#[derive(Serialize, Debug)]
+pub struct TypeDefinitionInterface {
+    documentation: Option<EcoString>,
+    deprecation: Deprecation,
+    /// The number of the type's type variables
+    parameters: usize,
+    /// A list of the names of its constructors
+    constructors: Vec<EcoString>,
+}
+
+/// Holds serialisable data about a top-level public value defined in a module
+/// (either a constant or a function)
+#[derive(Serialize, Debug)]
+pub struct ValueInterface {
```

Comment:
Could you include the supported targets too please 🙏 

Need to decide what's a good encoding for "all". Could be a null value means everything, or it could be that there's a "gleam" target which means all (`["gleam", "erlang"]`). Or something else.

Reply:
There's a `supported_targets` field in this struct; it is an array of targets so "all" would just become a list with all the targets like this: `["erlang", "javascript"]`

Reply:
Oh yes! Sorry I missed that.

It needs to distinguish between there being an external implementation for both and for there being a Gleam implementation as otherwise if a wasm backend arrives then we can no longer tell which are supported, and all packages will need to be republished. I think we'd need to extend the target tracking to have this level of information.

Reply:
Aha! You're right it'd be quite a problem if we don't track that... I'll have to put this on hold and update the target tracking code first, It shouldn't be too hard